app/version.go

```diff
@@ -1,40 +1,89 @@
 package app
 
-import "github.com/urfave/cli/v2"
+import (
+	"fmt"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/urfave/cli/v2"
+	"golang.org/x/mod/semver"
+)
 
 const (
-	DefaultVersion = "v0.13.0"
+	pseudoVersionMinLen        = len("vX.0.0-yyyymmddhhmmss-abcdefabcdef")
+	pseudoVersionCommitInfoLen = len("yyyymmddhhmmss-abcdefabcdef")
 )
 
 var (
-	BuildDate string
-	Commit    string
-	Version   string
+	MakeVersion string
 )
 
+type BuildInfo struct {
+	Commit     string
+	CommitTime time.Time
+	Version    string
+}
+
 func NewVersionCommand() (CommandOut, error) {
 	return CommandOut{Command: &cli.Command{
 		Name:    "version",
 		Usage:   "Version information",
 		Aliases: []string{"v"},
 		Action: func(c *cli.Context) error {
-			return PrintObj(&struct {
-				BuildDate string
-				Commit    string
-				Version   string
-			}{
-				BuildDate: BuildDate,
-				Commit:    Commit,
-				Version:   getVersion(),
-			})
+			return PrintObj(NewBuildInfo())
 		},
 	}}, nil
 }
 
-func getVersion() string {
-	version := Version
-	if len(version) == 0 {
-		version = DefaultVersion
+func NewBuildInfo() BuildInfo {
+	var info BuildInfo
+
+	di, ok := debug.ReadBuildInfo()
+	if !ok {
+		return info
 	}
-	return version
+
+	if semver.IsValid(di.Main.Version) {
+		info.Version = strings.Split(semver.Canonical(di.Main.Version), "-")[0]
+
+		// See https://go.dev/ref/mod#pseudo-versions for more info on the expected string format
+		// when the binary is compiled with go install. We always expect to hit this path if
+		// di.Main.Version is a valid semver.
+		if len(di.Main.Version) >= pseudoVersionMinLen {
+			commitInfoStart := len(di.Main.Version) - pseudoVersionCommitInfoLen
+			split := strings.Split(di.Main.Version[commitInfoStart:], "-")
+			info.Commit = split[1]
+
+			at, err := time.Parse("20060102150405", split[0])
+			if err == nil {
+				info.CommitTime = at
+			}
+
+			return info
+		}
+	} else {
+		info.Version = MakeVersion
+	}
+
+	// Developer build, pull extra info from vcs.
+	var hash, modified string
+
+	for _, setting := range di.Settings {
+		switch setting.Key {
+		case "vcs.revision":
+			hash = setting.Value
+		case "vcs.modified":
+			if v, err := strconv.ParseBool(setting.Value); err == nil && v {
+				modified = "-modified"
+			}
+		case "vcs.time":
+			info.CommitTime, _ = time.Parse(time.RFC3339, setting.Value)
+		}
+	}
```

Comment:
where do these setting values come from?

Reply:
The go toolchain populates them, and then lets you pull it programmatically for use-cases like this. You can also see it with their cli:
```
❯ go version -m ./tcld
./tcld: go1.21.0
	path	github.com/temporalio/tcld/cmd/tcld
	mod	github.com/temporalio/tcld	(devel)
	dep	github.com/cpuguy83/go-md2man/v2	v2.0.2	h1:p1EgwI/C7NhT0JmVkwCD2ZBK8j4aeHQX2pMHHBfMQ6w=
	dep	github.com/go-playground/locales	v0.14.1	h1:EWaQ/wswjilfKLTECiXz7Rh+3BjFhfDFKv/oXslEjJA=
	dep	github.com/go-playground/universal-translator	v0.18.1	h1:Bcnm0ZwsGyWbCzImXv+pAJnYK9S473LQFuzCbDbfSFY=
	dep	github.com/go-playground/validator/v10	v10.13.0	h1:cFRQdfaSMCOSfGCCLB20MHvuoHb/s5G8L5pu2ppK5AQ=
	dep	github.com/gogo/protobuf	v1.3.2	h1:Ov1cvc58UF3b5XjBnZv7+opcTcQFZebYjWzi34vdm4Q=
	dep	github.com/golang/mock	v1.6.0	h1:ErTB+efbowRARo13NNdxyJji2egdxLGQhRaY+DUumQc=
	dep	github.com/golang/protobuf	v1.5.2	h1:ROPKBNFfQgOUMifHyP+KYbvpjbdoFNs+aK7DXlji0Tw=
	dep	github.com/kylelemons/godebug	v1.1.0	h1:RPNrshWIDI6G2gRW9EHilWtl7Z6Sb1BR0xunSBf0SNc=
	dep	github.com/leodido/go-urn	v1.2.3	h1:6BE2vPT0lqoz3fmOesHZiaiFh7889ssCo2GMvLCfiuA=
	dep	github.com/russross/blackfriday/v2	v2.1.0	h1:JIOH55/0cWyOuilr9/qlrm0BSXldqnqwMsf35Ld67mk=
	dep	github.com/urfave/cli/v2	v2.25.3	h1:VJkt6wvEBOoSjPFQvOkv6iWIrsJyCrKGtCtxXWwmGeY=
	dep	github.com/xrash/smetrics	v0.0.0-20201216005158-039620a65673	h1:bAn7/zixMGCfxrRTfdpNzjtPYqr8smhKouy9mxVdGPU=
	dep	go.uber.org/atomic	v1.7.0	h1:ADUqmZGgLDDfbSL9ZmPxKTybcoEYHgpYfELNoN+7hsw=
	dep	go.uber.org/dig	v1.16.1	h1:+alNIBsl0qfY0j6epRubp/9obgtrObRAc5aD+6jbWY8=
	dep	go.uber.org/fx	v1.19.2	h1:SyFgYQFr1Wl0AYstE8vyYIzP4bFz2URrScjwC4cwUvY=
	dep	go.uber.org/multierr	v1.6.0	h1:y6IPFStTAIT5Ytl7/XYmHvzXQ7S3g/IeZW9hyZ5thw4=
	dep	go.uber.org/zap	v1.23.0	h1:OjGQ5KQDEUawVHxNwQgPpiypGHOxo2mNZsOqTak4fFY=
	dep	golang.org/x/crypto	v0.7.0	h1:AvwMYaRytfdeVt3u6mLaxYtErKYjxA2OXjJ1HHq6t3A=
	dep	golang.org/x/net	v0.8.0	h1:Zrh2ngAOFYneWTAIAPethzeaQLuHwhuBkuV6ZiRnUaQ=
	dep	golang.org/x/sys	v0.6.0	h1:MVltZSvRTcU2ljQOhs94SXPftV6DCNnZViHeQps87pQ=
	dep	golang.org/x/text	v0.8.0	h1:57P1ETyNKtuIjB4SRd15iJxuhj8Gc416Y78H3qgMh68=
	dep	google.golang.org/genproto	v0.0.0-20230110181048-76db0878b65f	h1:BWUVssLB0HVOSY78gIdvk1dTVYtT1y8SBWtPYuTJ/6w=
	dep	google.golang.org/grpc	v1.54.0	h1:EhTqbhiYeixwWQtAEZAxmV9MGqcjEU2mFx52xCzNyag=
	dep	google.golang.org/protobuf	v1.28.1	h1:d0NfwRgPtno5B1Wa6L2DAG+KivqkdutMf1UhdNx175w=
	build	-buildmode=exe
	build	-compiler=gc
	build	-ldflags="-X github.com/temporalio/tcld/app.date=2023-08-29T17:15:08-07:00 -X github.com/temporalio/tcld/app.commit=132e34e4954e -X github.com/temporalio/tcld/app.version=v0.13.1-embed-version-test"
	build	DefaultGODEBUG=panicnil=1
	build	CGO_ENABLED=1
	build	CGO_CFLAGS=
	build	CGO_CPPFLAGS=
	build	CGO_CXXFLAGS=
	build	CGO_LDFLAGS=
	build	GOARCH=arm64
	build	GOOS=darwin
	build	vcs=git
	build	vcs.revision=132e34e4954e8da3e5d47100068f8ad061df5f11
	build	vcs.time=2023-08-30T00:15:08Z
	build	vcs.modified=false
```